File: dotnet/Safe/SafeLogging.cs
namespace SastEval.Safe;

/// <summary>
/// True Negative for CWE-117: User input sanitized before logging.
/// SAST tools should NOT flag this as log injection.
/// </summary>
public static class SafeLogging
{
    public static IResult Handle(HttpContext context)
    {
        var username = context.Request.Query["username"].ToString();

        if (string.IsNullOrEmpty(username))
        {
            return Results.BadRequest("Missing 'username' query parameter");
        }

        // SAFE: Newlines and carriage returns stripped before logging
        var sanitized = username
            .Replace("\n", "")
            .Replace("\r", "")
            .Replace("\t", "");

        Console.WriteLine("Login attempt for: " + sanitized);

        return Results.Text("Login attempt logged");
    }
}

File: java/src/main/java/com/sasteval/safe/SafeLogging.java
package com.sasteval.safe;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.logging.Logger;

/**
 * TRUE NEGATIVE for CWE-117: Log Injection
 * SAFE: User input is sanitized (newlines and carriage returns removed) before logging.
 */
public class SafeLogging extends HttpServlet {

    private static final Logger logger = Logger.getLogger(SafeLogging.class.getName());

    private String sanitizeForLog(String input) {
        if (input == null) {
            return "null";
        }
        // SAFE: Strip newlines and carriage returns to prevent log injection
        return input.replace("\n", "").replace("\r", "").replace("\t", "");
    }

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {

        String username = request.getParameter("username");

        // SAFE: Sanitized input logged
        String sanitized = sanitizeForLog(username);
        logger.info("Login attempt for: " + sanitized);

        response.setContentType("application/json");
        response.getWriter().println("{\"status\": \"login attempt logged\"}");
    }
}
